hudi-common/src/main/java/org/apache/hudi/common/table/timeline/versioning/v1/CompletionTimeQueryViewV1.java

```diff
@@ -0,0 +1,239 @@
+    // Note: use getWriteTimeline() to keep sync with the fs view visibleCommitsAndCompactionTimeline, see AbstractTableFileSystemView.refreshTimeline.
+    this.firstNonSavepointCommit = metaClient.getActiveTimeline().getWriteTimeline().getFirstNonSavepointCommit().map(HoodieInstant::getRequestTime).orElse("");
+    load();
+  }
+
+  @Override
+  public boolean isCompleted(String beginInstantTime) {
+    // archival does not proceed beyond the first savepoint, so any instant before that is completed.
+    return this.beginToCompletionInstantTimeMap.containsKey(beginInstantTime) || isArchived(beginInstantTime);
+  }
+
+  @Override
+  public boolean isArchived(String instantTime) {
+    return InstantComparatorUtils.compareTimestamps(instantTime, LESSER_THAN, this.firstNonSavepointCommit);
+  }
+
+  @Override
+  public boolean isCompletedBefore(String baseInstant, String instantTime) {
+    Option<String> completionTimeOpt = getCompletionTime(baseInstant, instantTime);
+    if (completionTimeOpt.isPresent()) {
+      return InstantComparatorUtils.compareTimestamps(completionTimeOpt.get(), LESSER_THAN, baseInstant);
+    }
+    return false;
+  }
+
+  @Override
+  public boolean isSlicedAfterOrOn(String baseInstant, String instantTime) {
+    Option<String> completionTimeOpt = getCompletionTime(baseInstant, instantTime);
+    if (completionTimeOpt.isPresent()) {
+      return InstantComparatorUtils.compareTimestamps(completionTimeOpt.get(), GREATER_THAN_OR_EQUALS, baseInstant);
+    }
+    return true;
+  }
+
+  @Override
+  public Option<String> getCompletionTime(String baseInstant, String instantTime) {
+    Option<String> completionTimeOpt = getCompletionTime(instantTime);
+    if (completionTimeOpt.isPresent()) {
+      String completionTime = completionTimeOpt.get();
+      if (completionTime.length() != baseInstant.length()) {
+        // ==============================================================
+        // LEGACY CODE
+        // ==============================================================
+        // Fixes the completion time to reflect the completion sequence correctly
+        // if the file slice base instant time is not in datetime format.
+        // For example, many test cases just use integer string as the instant time.
+        // CAUTION: this fix only works for OCC(Optimistic Concurrency Control).
+        // for NB-CC(Non-blocking Concurrency Control), the file slicing may be incorrect.
+        return Option.of(instantTime);
+      }
+    }
+    return completionTimeOpt;
+  }
+
+  @Override
+  public Option<String> getCompletionTime(String beginTime) {
+    String completionTime = this.beginToCompletionInstantTimeMap.get(beginTime);
+    if (completionTime != null) {
+      return Option.of(completionTime);
+    }
+
+    // ***This is the key change between V1 and V2 completion time query-view***
```

Comment:
The modification time is also archived, can we just use that as the completion time and eliminate this duplicated class?

Reply:
querying v1 archived timeline will be a non-starter for any queries.. Thats the reason.